File: OnlineTaksiSwiftUI/OnlineTaksiSwiftUI/Features/Driver/DriverHomeViewModel.swift
//
//  DriverHomeViewModel.swift
//  OnlineTaksiSwiftUI
//
//  Created by Alperen Saraç on 23.04.2026.
//

import Foundation
import CoreLocation

@MainActor
final class DriverHomeViewModel: NSObject, ObservableObject {
    @Published var isOnline: Bool = false
    @Published var currentLatText: String = "-"
    @Published var currentLngText: String = "-"
    @Published var availableRides: [AvailableRideItem] = []
    @Published var activeRide: RideResponse?
    @Published var lastLog: String = "Hazır"
    @Published var message: String?
    @Published var isLoadingAvailableRides: Bool = false
    @Published var isAcceptingRide: Bool = false

    private let sessionManager: SessionManager
    private let driverRepository: DriverRepository
    private let socketManager: SocketManager
    private let locationManager = CLLocationManager()

    init(
        sessionManager: SessionManager,
        driverRepository: DriverRepository,
        socketManager: SocketManager = SocketManager()
    ) {
        self.sessionManager = sessionManager
        self.driverRepository = driverRepository
        self.socketManager = socketManager
        super.init()

        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyBest
        configureSocketCallbacks()
    }

    func connectSocket() {
        guard let token = sessionManager.token, !token.isEmpty else {
            message = "Token yok"
            lastLog = "Token yok"
            return
        }

        socketManager.connect(token: token)
        lastLog = "Socket bağlantısı başlatıldı"
    }

    func disconnectSocket() {
        socketManager.disconnect()
        lastLog = "Socket bağlantısı kapatıldı"
    }

    func setOnline(_ online: Bool) async {
        do {
            let response = try await driverRepository.setOnline(online)
            isOnline = response.is_online
            lastLog = online ? "Online oldun" : "Offline oldun"
            message = online ? "Online oldun" : "Offline oldun"
        } catch {
            message = error.localizedDescription
            lastLog = error.localizedDescription
        }
    }

    func loadAvailableRides() async {
        isLoadingAvailableRides = true

        do {
            let response = try await driverRepository.getAvailableRides()
            availableRides = response.rides
            lastLog = "Açık ride listesi güncellendi"
        } catch {
            message = error.localizedDescription
            lastLog = error.localizedDescription
        }

        isLoadingAvailableRides = false
    }

    func loadActiveRide() async {
        do {
            let response = try await driverRepository.getMyActiveRides()
            activeRide = response.rides.first
            lastLog = activeRide == nil ? "Aktif ride yok" : "Aktif ride güncellendi"
        } catch {
            message = error.localizedDescription
            lastLog = error.localizedDescription
        }
    }

    func acceptRide(_ rideId: Int) async {
        isAcceptingRide = true

        do {
            let response = try await driverRepository.acceptRide(rideId: rideId)
            activeRide = response
            availableRides.removeAll { $0.id == rideId }
            lastLog = "Ride kabul edildi. id=\(rideId)"
            message = "Ride kabul edildi"
        } catch {
            message = error.localizedDescription
            lastLog = error.localizedDescription
        }

        isAcceptingRide = false
    }

    func updateRideStatus(_ status: String, note: String? = nil) async {
        guard let activeRide else {
            message = "Aktif ride yok"
            lastLog = "Aktif ride yok"
            return
        }

        do {
            let response = try await driverRepository.updateRideStatus(
                rideId: activeRide.id,
                status: status,
                note: note
            )
            self.activeRide = response
            lastLog = "Status güncellendi: \(status)"
            message = "Status güncellendi: \(status)"
        } catch {
            message = error.localizedDescription
            lastLog = error.localizedDescription
        }
    }

    func startLocationTracking() {
        let status = locationManager.authorizationStatus

        switch status {
        case .notDetermined:
            locationManager.requestWhenInUseAuthorization()
        case .authorizedWhenInUse, .authorizedAlways:
            locationManager.startUpdatingLocation()
            lastLog = "Konum güncellemeleri başlatıldı"
        default:
            message = "Konum izni gerekli"
            lastLog = "Konum izni gerekli"
        }
    }

    func stopLocationTracking() {
        locationManager.stopUpdatingLocation()
        lastLog = "Konum güncellemeleri durduruldu"
    }

    func clearMessage() {
        message = nil
    }

    private func configureSocketCallbacks() {
        socketManager.onMessageReceived = { [weak self] message in
            guard let self else { return }

            self.lastLog = message

            guard let eventName = SocketEventParser.getEventName(message) else { return }

            if eventName == "NEW_RIDE_REQUEST",
               let ride = self.parseIncomingRide(message) {
                let exists = self.availableRides.contains { $0.id == ride.id }
                if !exists {
                    self.availableRides.insert(ride, at: 0)
                    self.message = "Yeni ride geldi"
                    self.lastLog = "Yeni ride geldi"
                }
            }
        }

        socketManager.onError = { [weak self] errorText in
            guard let self else { return }
            self.message = errorText
            self.lastLog = errorText
        }
    }

    private func parseIncomingRide(_ message: String) -> AvailableRideItem? {
        guard let data = message.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let event = json["event"] as? String,
              event == "NEW_RIDE_REQUEST",
              let body = json["data"] as? [String: Any],
              let rideId = body["ride_id"] as? Int,
              let customerId = body["customer_id"] as? Int,
              let pickupLat = body["pickup_lat"] as? Double,
              let pickupLng = body["pickup_lng"] as? Double,
              let pickupAddress = body["pickup_address"] as? String,
              let dropoffLat = body["dropoff_lat"] as? Double,
              let dropoffLng = body["dropoff_lng"] as? Double,
              let dropoffAddress = body["dropoff_address"] as? String,
              let status = body["status"] as? String else {
            return nil
        }

        return AvailableRideItem(
            id: rideId,
            customer_id: customerId,
            pickup_lat: pickupLat,
            pickup_lng: pickupLng,
            pickup_address: pickupAddress,
            dropoff_lat: dropoffLat,
            dropoff_lng: dropoffLng,
            dropoff_address: dropoffAddress,
            status: status,
            estimated_fare: body["estimated_fare"] as? Double
        )
    }
}
extension DriverHomeViewModel: CLLocationManagerDelegate {
    nonisolated func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        Task { @MainActor in
            let status = manager.authorizationStatus
            switch status {
            case .authorizedAlways, .authorizedWhenInUse:
                manager.startUpdatingLocation()
                self.lastLog = "Konum izni verildi"
            case .denied, .restricted:
                self.message = "Konum izni verilmedi"
                self.lastLog = "Konum izni verilmedi"
            default:
                break
            }
        }
    }

    nonisolated func locationManager(
        _ manager: CLLocationManager,
        didUpdateLocations locations: [CLLocation]
    ) {
        guard let location = locations.last else { return }

        Task { @MainActor in
            self.currentLatText = String(location.coordinate.latitude)
            self.currentLngText = String(location.coordinate.longitude)

            do {
                _ = try await self.driverRepository.updateLocation(
                    lat: location.coordinate.latitude,
                    lng: location.coordinate.longitude
                )
                self.lastLog = "Konum gönderildi"
            } catch {
                self.message = error.localizedDescription
                self.lastLog = error.localizedDescription
            }
        }
    }

    nonisolated func locationManager(
        _ manager: CLLocationManager,
        didFailWithError error: Error
    ) {
        Task { @MainActor in
            self.message = error.localizedDescription
            self.lastLog = error.localizedDescription
        }
    }
}

File: OnlineTaksiJetpack/app/src/main/java/com/example/onlinetaksijetpack/ui/driver/DriverHomeViewModel.kt
package com.example.onlinetaksijetpack.ui.driver

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.onlinetaksijetpack.data.remote.model.AvailableRideItem
import com.example.onlinetaksijetpack.data.remote.model.RideResponse
import com.example.onlinetaksijetpack.data.remote.socket.SocketManager
import com.example.onlinetaksijetpack.data.repository.DriverRepository
import com.example.onlinetaksijetpack.util.Resource
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import org.json.JSONObject

class DriverHomeViewModel(
    private val driverRepository: DriverRepository
) : ViewModel(), SocketManager.SocketEventListener {

    private val _uiState = MutableStateFlow(DriverHomeUiState())
    val uiState: StateFlow<DriverHomeUiState> = _uiState

    fun connectSocket(token: String?) {
        if (token.isNullOrBlank()) {
            _uiState.value = _uiState.value.copy(message = "Token yok")
            return
        }

        SocketManager.setListener(this)
        SocketManager.connect(token)
    }

    fun setOnline(isOnline: Boolean) {
        viewModelScope.launch {
            when (val result = driverRepository.setOnline(isOnline)) {
                is Resource.Success -> {
                    _uiState.value = _uiState.value.copy(
                        isOnline = result.data.is_online,
                        lastLog = if (isOnline) "Online oldun" else "Offline oldun",
                        message = if (isOnline) "Online oldun" else "Offline oldun"
                    )
                }

                is Resource.Error -> {
                    _uiState.value = _uiState.value.copy(
                        message = result.message,
                        lastLog = result.message
                    )
                }

                else -> {}
            }
        }
    }

    fun loadAvailableRides() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoadingAvailableRides = true)

            when (val result = driverRepository.getAvailableRides()) {
                is Resource.Success -> {
                    _uiState.value = _uiState.value.copy(
                        isLoadingAvailableRides = false,
                        availableRides = result.data.rides,
                        lastLog = "Açık ride listesi güncellendi"
                    )
                }

                is Resource.Error -> {
                    _uiState.value = _uiState.value.copy(
                        isLoadingAvailableRides = false,
                        message = result.message,
                        lastLog = result.message
                    )
                }

                else -> {}
            }
        }
    }

    fun loadActiveRide() {
        viewModelScope.launch {
            when (val result = driverRepository.getMyActiveRides()) {
                is Resource.Success -> {
                    _uiState.value = _uiState.value.copy(
                        activeRide = result.data.rides.firstOrNull(),
                        lastLog = "Aktif ride güncellendi"
                    )
                }

                is Resource.Error -> {
                    _uiState.value = _uiState.value.copy(
                        message = result.message,
                        lastLog = result.message
                    )
                }

                else -> {}
            }
        }
    }

    fun acceptRide(rideId: Int) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isAcceptingRide = true)

            when (val result = driverRepository.acceptRide(rideId)) {
                is Resource.Success -> {
                    _uiState.value = _uiState.value.copy(
                        isAcceptingRide = false,
                        activeRide = result.data,
                        availableRides = _uiState.value.availableRides.filterNot { it.id == rideId },
                        message = "Ride kabul edildi",
                        lastLog = "Ride kabul edildi. id=$rideId"
                    )
                }

                is Resource.Error -> {
                    _uiState.value = _uiState.value.copy(
                        isAcceptingRide = false,
                        message = result.message,
                        lastLog = result.message
                    )
                }

                else -> {}
            }
        }
    }

    fun updateRideStatus(status: String, note: String? = null) {
        val activeRide = _uiState.value.activeRide ?: run {
            _uiState.value = _uiState.value.copy(message = "Aktif ride yok")
            return
        }

        viewModelScope.launch {
            when (val result = driverRepository.updateRideStatus(activeRide.id, status, note)) {
                is Resource.Success -> {
                    _uiState.value = _uiState.value.copy(
                        activeRide = result.data,
                        lastLog = "Status güncellendi: $status",
                        message = "Status güncellendi: $status"
                    )
                }

                is Resource.Error -> {
                    _uiState.value = _uiState.value.copy(
                        message = result.message,
                        lastLog = result.message
                    )
                }

                else -> {}
            }
        }
    }

    fun sendLocation(lat: Double, lng: Double) {
        _uiState.value = _uiState.value.copy(
            currentLat = lat.toString(),
            currentLng = lng.toString()
        )

        viewModelScope.launch {
            driverRepository.updateLocation(lat, lng)
        }
    }

    fun addIncomingRide(item: AvailableRideItem) {
        val exists = _uiState.value.availableRides.any { it.id == item.id }
        if (!exists) {
            _uiState.value = _uiState.value.copy(
                availableRides = listOf(item) + _uiState.value.availableRides,
                lastLog = "Yeni ride geldi",
                message = "Yeni ride geldi"
            )
        }
    }

    fun clearMessage() {
        _uiState.value = _uiState.value.copy(message = null)
    }

    override fun onConnected() {
        _uiState.value = _uiState.value.copy(lastLog = "Socket bağlandı")
    }

    override fun onDisconnected() {
        _uiState.value = _uiState.value.copy(lastLog = "Socket bağlantısı kapandı")
    }

    override fun onMessage(message: String) {
        _uiState.value = _uiState.value.copy(lastLog = message)

        try {
            val json = JSONObject(message)
            val event = json.optString("event")

            if (event == "NEW_RIDE_REQUEST") {
                val data = json.optJSONObject("data")
                if (data != null) {
                    val incomingRide = AvailableRideItem(
                        id = data.optInt("ride_id"),
                        customer_id = data.optInt("customer_id"),
                        pickup_lat = data.optDouble("pickup_lat"),
                        pickup_lng = data.optDouble("pickup_lng"),
                        pickup_address = data.optString("pickup_address"),
                        dropoff_lat = data.optDouble("dropoff_lat"),
                        dropoff_lng = data.optDouble("dropoff_lng"),
                        dropoff_address = data.optString("dropoff_address"),
                        status = data.optString("status"),
                        estimated_fare = if (data.has("estimated_fare")) data.optDouble("estimated_fare") else null
                    )

                    addIncomingRide(incomingRide)
                }
            }
        } catch (e: Exception) {
            _uiState.value = _uiState.value.copy(lastLog = "Socket parse hatası: ${e.message}")
        }
    }

    override fun onError(errorMessage: String) {
        _uiState.value = _uiState.value.copy(
            message = errorMessage,
            lastLog = errorMessage
        )
    }

    override fun onCleared() {
        super.onCleared()
        SocketManager.setListener(null)
    }
}